### PR TITLE
Add WordPress JSDoc ESLint configuration

### PR DESCRIPTION
**Pull request by atimmer**
## Description
The WordPress ESLint configuration plugin has a JSDoc configuration. I've added this to the ESLint configuration to see how much would fail. And start the discussion about how to integrate this.

Current stats:
```
✖ 1504 problems (869 errors, 635 warnings)
```

### CODE REVIEW

**Comment by youknowriad** [comment]
Maybe we should enable this as a "warning" to start fixing those as we add/update code?

**Comment by atimmer** [comment]
This is ready to be merged.

File: .eslines.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    "branches": {
+        "default": ["downgrade-unmodified-lines"]
+    },
+    "processors": {
+        "downgrade-unmodified-lines": {
+            "remote": "origin/master",
+            "rulesNotToDowngrade": ["no-unused-vars"]
```
Comment (youknowriad) [inline .eslines.json]: Why this rule is different?

**Review by youknowriad — APPROVED** [review]
This is working pretty well for me. looking forward getting rid of all these warnings.

File: .eslines.json (hunk 1 above)
Comment (ntwb) [reply .eslines.json]: Via https://www.npmjs.com/package/eslines#how-to-use-it:
> _"The `no-unused-vars` rule won't be downgraded, though - this is one case where changing one line can cause a linting error in a different one so we recommend preventing it from downgrading."_

**Comment by ntwb** [comment]
In https://github.com/WordPress/gutenberg/pull/4245/commits/e352c976107e39a433a92ac455ba9b2c2fbc9f13 I updated to a more recent commit of `eslint-plugin-wordpress`, there were no JSDoc rules changes between this and the previous version used, just this version is the latest is all

**Review by ntwb — APPROVED** [review]
This is great, thanks @atimmer 👍

**Comment by noisysocks** [comment]
This implementation shows errors for lines that differ from `origin/master` and warnings for everything else. This means that Travis CI will fail on a PR until jsdoc comments are added to any methods that that PR touches.

Is that intentional? I read @youknowriad's [comment](https://github.com/WordPress/gutenberg/pull/4245#issuecomment-355009789) as that we only ever want warnings.

**Comment by atimmer** [comment]
@noisysocks That comment was before `eslines` was introduced. The idea was always for these to be errors, but it is hard to enable this for the entire codebase without first cleaning up everything. `eslines` solves this problem. It is basically the boy-scout rule in a tool.

**Comment by noisysocks** [comment]
Cool, just thought I'd check since it wasn't explicitly mentioned in the PR description! 😄

**Comment by mcsf** [comment]
I'm not convinced this rule suits the project well, at least in its current form. Even with `eslines`, it requires a lot of attention, arguably too much. Examples:

- **Functional components**, incl. small ones. The JSDoc will be a repetition of what is already obvious in the code:

```js
/**
 * Returns a menu of tools representing actions on the editor's content.
 *
 * @returns {WPElement} Element.
 */
export default function EditorActions() {
  return (
    <MenuItemsGroup className="editor-actions" label={ __( 'Tools' ) }>
      <CopyContentButton />
    </MenuItemsGroup>
  );
}
```

Assuming the component takes props, that will require an additional ` * @param   {Object} props  React props passed to the component.`, though the tool can't require that every possible prop be described.

Large or otherwise significant or hard-to-grasp components will obviously warrant JSDoc, but one of the strengths of our compositional paradigm is that it encourages breaking down functionality in layers and spawning small components. If `Foo` renders a small collection of `Bars`, it's preferable to add a `function Bar() {}` to use in a `Array#map` than to inline everything in `Foo#render` or to delegate to `Foo#renderBars`. `require-jsdoc` works against this.

- **Component methods**. Case in point:

image

`CopyContentButton` is a small component, as it's basically a wrapper for `ClipboardButton` with some local state and some Redux state relaying. Of the four methods that the class comprises, none of them should even have a JSDoc—let alone be _required_ one. Two of the methods—`constructor` and `render`—are part of the React standard, and the two remaining ones are state-setting one-liners.

* * *

I think the main point of this is that, whenever we're talking about _components_, we're dealing with a standard interface:
- Props go in, element goes out.
- Props can be optional, they can have default values, and this should be visible in the function signature or in the render method's `const { foo, bar = 'BAR' } = this.props;` assignment anyway.
- Methods _tend_ to fit one of a small set of types: lifecycle methods (incl. `constructor` and `render`), state changers or similar procedures (e.g. callers of bound action creators), or other tiny helpers. Most of the time these don't need comments—IMO, they detract from the code and add a bit of inertia. For any other method, comments should be considered as a general rule—something that, granted, isn't easy to codify in a linter.

With that premise that it is a standard interface, it follows that it's wasteful to document said interface every time it's implemented.

A trivial but potentially naïve way forward would be to give components a free pass by excluding any function declaration whose identifier has an uppercase letter as the leading character. For class components, perhaps exclude classes defining a `render` method.

**Comment by youknowriad** [comment]
👍  to @mcsf  concerns, I've already shared similar concerns in #core-js meetings but now facing them in reality shows that it's really too much for a standard interface.

**Comment by jdevalk** [comment]
This is simply laziness, I'm sorry. It's obvious what is in the code if you're well versed in the code, if you're not, this:

```
<MenuItemsGroup className="editor-actions" label={ __( 'Tools' ) }>
      <CopyContentButton />
</MenuItemsGroup>
```

Is not necessarily self-explanatory. Spending two to three seconds to write that it returns a copy content button really does help.

**Comment by afercia** [comment]
In a large open source project like WordPress, good documentation is key. Maybe in smaller projects or in projects internal to a company, the documentation effort may have a different priority and may adapt to specific momentary needs. In WordPress, I'd strongly suggest to not repeat the same mistakes done in the past and document _everything_.

I'd suggest to try to not think as an expert developer looking at code that was coded _now_ by yourself or your colleagues. Instead, I'd recommend to make an effort to think at someone with a low level of expertise who, in 2 years from now, reads for the first time your code. Expecting that they understand what is `obvious` for you is just an assumption.

**Comment by mcsf** [comment]
Aside, what are your thoughts on method-level enforcement?

**Comment by iseulde** [comment]
Just opened #4504, before I was aware of this conversation. Please let's remove at least the requirement for JSDocs on on React component methods. 🙂

**Comment by gziolo** [comment]
> My preference is that we require JSDoc comments on only the contract that a module or component has with its outside world. That is:
> 
> * Any functions, variables, classes or constants exported using export
> * Any props that a React component accepts

@noisysocks, this sounds like a good default for Eslint verification. It would still allow to auto-generate docs for all code that is consumed by contributors in most of the cases.

@mtias I'm wondering how this new rule will impact README files. They were often used to present API methods, which duplicates what JSDoc does. Should we standardize that and create scripts to generate those sections? Actually, this is how things are maintained in WP-CLI. @schlessera or @danielbachhuber can give us more details on that.

**Comment by ntwb** [comment]
I've now partially reverted this PR in #4510, some eyes on #4506 would be great, it's a `require-jsdoc` drop in ESLint rule that allows us to not require JSDocs on _methods_ 😄

**Comment by ntwb** [comment]
@aduth Fixed most instances (I hope) of these in #4653

**Comment by gziolo** [comment]
@ntwb, we can remove `eslines`, too. It was only necessary because we had all those rules set as warnings. What `eslines` does is it promotes warnings to errors for the modified lines.

**Comment by aduth** [comment]
In case there are others like me using the [DocBlockr Sublime Text plugin](https://github.com/spadgos/sublime-jsdocs), here's a user configuration which can be added to adhere to the updated documentation styling:

_Preferences > Package Settings > DocBlockr > Settings - User_

```json
{
  "jsdocs_return_tag": "@returns",
  "jsdocs_spacer_between_sections": true,
  "jsdocs_per_section_indent": true,
  "jsdocs_lower_case_primitives": true
}
```

**Comment by iseulde** [comment]
Thanks @aduth!

**Comment by schlessera** [comment]
Re. generated documentation like referenced above, you generally do this for everything that can be considered an "interface". So, you have the public interface, which is what external developers will use to create third-party blocks or integrations, and then you have the internal interface which deals with the intra-communication between your individual components/modules. Ideally, both of these categories should be documented in the code without fail and the external documentation should be generated out of this source code. This is the only reliable way of keeping the documentation in sync with the actual code.

As to the usefulness of the docblocks for simplistic methods...

You would not only add a short description of what and why (the why normally being something that is not obvious from the code alone), but also historical information that can inform future decisions... Knowing when something was introduced will help with debugging version-dependent bugs and be smart about deprecations. Knowing what discourse led to the current implementation avoids switching between two good-enough approaches every few months. There is a lot of potential information that you can put into such a docblock that will not necessarily help you understand the one or two lines of code, but help you make better decisions two years down the road.

**Pull request merged**